torch/_prims/__init__.py

```diff
@@ -0,0 +1,566 @@
+  assert outer_length >= 0
+
+  # Verifies the dim can be split with the specified lhs_length
+  inner_length = a.shape[dim] / outer_length
+  assert int(inner_length) == inner_length
+
+def _split_dim_aten(a: Tensor, dim: int, outer_length: int) -> Tensor:
+  inner_length = int(a.shape[dim] / outer_length)
+  new_shape = a.shape[0:dim] + (outer_length, inner_length) + a.shape[dim + 1:]
+
+  return a.view(new_shape)
+
+def split_dim(a: Tensor, dim: int, outer_length: int) -> Tensor:
+  '''
+  Creates a view of a with the given dimension (of length l) split
+  into two dimensions, with the outer of the two having
+  length outer_length and the inner of the two having computed
+  length inner_length such outer_length * inner_length = l.
+  '''
+
+  _split_dim_checks(a, dim, outer_length)
+  return _split_dim_aten(a, dim, outer_length)
+
+# Note: allows dimensions to be specified redundantly
+def _squeeze_checks(a: Tensor, dimensions: Sequence):
+  assert isinstance(a, Tensor)
+
+  for idx in dimensions:
+    _validate_idx(a.shape, idx)
+    assert a.shape[idx] == 1
+
+def _squeeze_aten(a: Tensor, dimensions: Sequence) -> Tensor:
+  for idx in dimensions:
+    a = torch.squeeze(a, dim=idx)
+
+  return a
+
+def squeeze(a: Tensor, dimensions: Sequence) -> Tensor:
+  '''
+  Creates a view of a with the specified dimensions of
+  length one removed.
+  '''
+
+  _squeeze_checks(a, dimensions)
+  return _squeeze_aten(a, dimensions)
+
+#
+# Shape operations
+#
+def collapse(a: Tensor, start: int, end: int) -> Tensor:
+  '''
+  Wrapper around reshape that collapses a span of dimensions.
+
+  See merge_dims for the corresponding view operation.
+  '''
+
+  dim_length = 1
+  for idx in range(start, end):
+    dim_length = dim_length * a.shape[idx]
+
+  new_shape = a.shape[0:start] + (dim_length,) + a.shape[end:]
+  return reshape(a, new_shape)
+
+def _concatenate_checks(tensors: Sequence[Tensor], dim: int):
+  assert len(tensors) > 0
+  assert dim >= 0
+
+  shape = tensors[0].shape
+  assert dim < len(shape)
+
+  _check_same_dtype(*tensors)
+
+  # Verifies same shape (except in the concat dimension)
+  for tensor in tensors:
+    for idx, lengths in enumerate(zip(shape, tensor.shape)):
+      common_length, length = lengths
+      if idx == dim:
+        continue
+      assert length == common_length
+
+def _concatentate_aten(tensors: Sequence[Tensor], dim: int) -> Tensor:
+  return torch.cat(tensors, dim)
+
+def concatenate(tensors: Sequence[Tensor], dim: int) -> Tensor:
+  _concatenate_checks(tensors, dim)
+  return _concatentate_aten(tensors, dim)
+
+def _reshape_checks(a: Tensor, shape: Sequence):
+  assert isinstance(a, Tensor)
+  _validate_shape(shape)
+
+  # Validates the tensor and the requested shape have the
+  # same number of elements
+  numel = reduce(lambda acc, x: acc * x, shape)
+  assert a.numel() == numel
+
+def _reshape_aten(a: Tensor, shape: Sequence) -> Tensor:
+
+  return a.clone().reshape(shape).contiguous()
```

Comment:
You're doing (potentially) two copies here. Consider doing the following, if we can afford to have access to `.storage()`.
```suggestion
  x = a.reshape(shape).contiguous()
  is_view = a.storage().data_ptr() == x.storage().data_ptr()
  return x.clone() if is_view else x
```

Reply:
`storage` and `data_ptr` are not good things to rely on, not all tensor subclasses and/or backends will have it. We may have to go with a potentially extra copy

Reply:
It's odd that we don't have an is_alias_of in the python API...